plugins/useractivity/useractivity.go

```diff
@@ -103,22 +110,35 @@ func (r *reporter) submitter() {
 		// channel, and the loop body contains a Remove call (which acquires the
 		// shard's mutex), we deadlock.
 		//
+		tickTime := time.Now()
 		toRemove := []string{}
 		for t := range r.sessions.Iter() {
-			s := t.Val.(session)
-			// Session ran out of the window?
-			if s.lastAct.Add(r.window).Before(time.Now()) {
-				toRemove = append(toRemove, t.Key)
-			}
+			func() {
+				s := t.Val.(session)
+				// Session ran out of the window?
+				s.mut.Lock()
+				defer s.mut.Unlock()
+				if s.lastAct.Add(r.window).Before(tickTime) {
+					toRemove = append(toRemove, t.Key)
+				}
+			}()
 		}
 
 		for _, key := range toRemove {
-			val, exists := r.sessions.Get(key)
-			if !exists {
-				logging.Errorf("Atempting to delete a nonexisting session for %s", key)
-			}
-			r.sessions.Remove(key)
-			r.submit(key, val.(session))
+			func() {
+				val, exists := r.sessions.Get(key)
+				if !exists {
+					logging.Errorf("Attempting to delete a nonexisting session for %s", key)
+					return
+				}
+				s := val.(session)
+				s.mut.Lock()
+				defer s.mut.Unlock()
+				if s.lastAct.Add(r.window).Before(tickTime) {
```

Comment:
BTW in the versions of this code that use `session`, this check is redundant with the one at 121, since s.lastAct is a property of the local variable & can't change.

Yes, this means we formerly had an intentional race: we would end & submit a session even if an action arrived in between a session being added to the `toRemove` list and us processing the removal. I discussed this with either Andrew or Max or both, & we concluded that we were OK with losing a user action in this case, because activity sessions are basically a best-effort feature; after all, sessions are heuristically inferred anyway.

Anyway, if we start storing `*session` in the map, s.lastAct will be a property of the shared object, so this comment is moot. And we still want to grab a mutex for the `submit` call.